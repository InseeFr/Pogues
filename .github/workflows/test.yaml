name: Pogues Test CI

on:
  push:
    branches: ['main', 'next']
    paths-ignore: ['next/**']
  pull_request:
    types: [opened, synchronize, reopened]
    paths-ignore: ['next/**']

jobs:
  test:
    name: Test
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - uses: actions/setup-node@v4
        with:
          node-version: 20
      - run: yarn
      # Run tests sonar reports
      - run: yarn test:coverage
      - name: Upload coverage artifact
        uses: actions/upload-artifact@v4
        with:
          name: coverage
          path: coverage

  sonarcloud:
    name: SonarCloud
    runs-on: ubuntu-latest
    needs: test
    steps:
      - uses: actions/checkout@v4
        with:
          fetch-depth: 0
      - name: Download coverage artifact
        uses: actions/download-artifact@v4
        with:
          name: coverage
          path: coverage
      - name: SonarCloud Scan
        uses: SonarSource/sonarqube-scan-action@v4
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
          SONAR_TOKEN: ${{ secrets.SONAR_TOKEN }}
