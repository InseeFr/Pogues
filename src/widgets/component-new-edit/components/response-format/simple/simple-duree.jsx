import React from 'react';

import { connect } from 'react-redux';
import { Field, FormSection, formValueSelector } from 'redux-form';
import { required as funcRequired } from 'redux-form-validators';

import { DATATYPE_NAME } from '../../../../../constants/pogues-constants';
import GenericOption from '../../../../../forms/controls/generic-option';
import Input from '../../../../../forms/controls/input';
import Select from '../../../../../forms/controls/select';
import Dictionary from '../../../../../utils/dictionary/dictionary';

const { DURATION } = DATATYPE_NAME;

function ResponseFormatDatatypeDuree({
  format,
  type,
  formatTableList,
  formatTable,
  name,
  required,
  readOnly,
}) {
  const formatInit = type === 'TABLE' ? formatTableList || formatTable : format;
  const isDuration = formatInit === 'PTnHnM' || formatInit === 'PnYnM';

  return (
    <FormSection name={name}>
      <div className="response-format-datatype-duree">
        <Field
          name="format"
          id="date_format"
          component={Select}
          label={Dictionary.date_format}
          required={required}
          disabled={readOnly}
          validate={funcRequired({ message: Dictionary.validationRequired })}
        >
          <GenericOption key="" value="">
            {Dictionary.durationinitial}
          </GenericOption>

          <GenericOption key="PTnHnM" value="PTnHnM">
            {Dictionary.durationformat1}
          </GenericOption>

          <GenericOption key="PnYnM" value="PnYnM">
            {Dictionary.durationformat2}
          </GenericOption>
        </Field>
        {isDuration && (
          <div>
            <div className="response-format-datatype-duree-minimum">
              <div className="response-format-datatype-duree-label-minimum">
                {Dictionary.minimum}
              </div>
              <div style={{ width: '75%' }}>
                {formatInit === 'PTnHnM' && (
                  <div className="response-format-datatype-duree-minimum">
                    <Field
                      name="mihours"
                      type="number"
                      step="any"
                      component={Input}
                      label={Dictionary.hours}
                      disabled={readOnly}
                      required
                      validate={funcRequired({
                        message: Dictionary.validationRequired,
                      })}
                    />
                    <Field
                      name="miminutes"
                      type="number"
                      step="any"
                      component={Input}
                      label={Dictionary.minutes}
                      disabled={readOnly}
                      required
                      validate={funcRequired({
                        message: Dictionary.validationRequired,
                      })}
                    />
                  </div>
                )}
                {formatInit === 'PnYnM' && (
                  <div className="response-format-datatype-duree-minimum">
                    <Field
                      name="miyears"
                      type="number"
                      step="any"
                      component={Input}
                      label={Dictionary.years}
                      disabled={readOnly}
                      required
                      validate={funcRequired({
                        message: Dictionary.validationRequired,
                      })}
                    />
                    <Field
                      name="mimonths"
                      type="number"
                      step="any"
                      component={Input}
                      label={Dictionary.months}
                      disabled={readOnly}
                      required
                      validate={funcRequired({
                        message: Dictionary.validationRequired,
                      })}
                    />
                  </div>
                )}
              </div>
            </div>

            <div className="response-format-datatype-duree-maximum">
              <div className="response-format-datatype-duree-label-maximum">
                {Dictionary.maximum}
              </div>
              <div style={{ width: '75%' }}>
                {formatInit === 'PTnHnM' && (
                  <div className="response-format-datatype-duree-maximum">
                    <Field
                      name="mahours"
                      type="number"
                      step="any"
                      component={Input}
                      label={Dictionary.hours}
                      disabled={readOnly}
                      required
                      validate={funcRequired({
                        message: Dictionary.validationRequired,
                      })}
                    />
                    <Field
                      name="maminutes"
                      type="number"
                      step="any"
                      component={Input}
                      label={Dictionary.minutes}
                      disabled={readOnly}
                      required
                      validate={funcRequired({
                        message: Dictionary.validationRequired,
                      })}
                    />
                  </div>
                )}
                {formatInit === 'PnYnM' && (
                  <div className="response-format-datatype-duree-maximum">
                    <Field
                      name="mayears"
                      type="number"
                      step="any"
                      component={Input}
                      label={Dictionary.years}
                      disabled={readOnly}
                      required
                      validate={funcRequired({
                        message: Dictionary.validationRequired,
                      })}
                    />
                    <Field
                      name="mamonths"
                      type="number"
                      step="any"
                      component={Input}
                      label={Dictionary.months}
                      disabled={readOnly}
                      required
                      validate={funcRequired({
                        message: Dictionary.validationRequired,
                      })}
                    />
                  </div>
                )}
              </div>
            </div>
          </div>
        )}
      </div>
    </FormSection>
  );
}

ResponseFormatDatatypeDuree.defaultProps = {
  name: DURATION,
  readOnly: false,
  required: true,
};

const mapStateToProps = (state) => {
  const selector = formValueSelector('component');
  return {
    formatTable: selector(
      state,
      'responseFormat.TABLE.LIST_MEASURE.SIMPLE.DURATION.format',
    ),
    formatTableList: selector(
      state,
      'responseFormat.TABLE.MEASURE.SIMPLE.DURATION.format',
    ),
    format: selector(state, 'responseFormat.SIMPLE.DURATION.format'),
    type: selector(state, 'responseFormat.type'),
  };
};

export default connect(mapStateToProps)(ResponseFormatDatatypeDuree);
