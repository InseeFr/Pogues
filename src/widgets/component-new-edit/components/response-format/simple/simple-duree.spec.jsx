import { render } from '@testing-library/react';
import { describe, expect, test } from 'vitest';

import { decoreFormField } from '../../../../../utils/test/test-utils';
import { ResponseFormatDatatypeDuree } from './simple-duree';

describe('Visualize Dropdown Component: ', () => {
  test('Should return the right HTML according to format "PTnHnM" ', () => {
    const props = {
      componentId: 'component-id',
      format: 'PTnHnM',
    };
    const { container } = render(
      decoreFormField(<ResponseFormatDatatypeDuree {...props} />),
    );
    expect(container).toMatchSnapshot();
  });
  test('Should return the right HTML according to format "PnYnM" ', () => {
    const props = {
      componentId: 'component-id',
      format: 'PnYnM',
    };
    const { container } = render(
      decoreFormField(<ResponseFormatDatatypeDuree {...props} />),
    );
    expect(container).toMatchSnapshot();
  });
});
