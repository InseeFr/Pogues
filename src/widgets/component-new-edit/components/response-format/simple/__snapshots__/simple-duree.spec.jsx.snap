// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Visualize Dropdown Component:  > Should return the right HTML 1`] = `
<div>
  <div
    class="response-format-datatype-duree"
  >
    <div
      class="ctrl-select"
    >
      <label
        for="select-DURATION.format"
      >
        Format
        <span
          class="ctrl-required"
        >
          *
        </span>
      </label>
      <div>
        <select
          id="select-DURATION.format"
          name="DURATION.format"
          placeholder="Format"
        >
          <option
            value=""
          >
            Duration format
          </option>
          <option
            value="PTnHnM"
          >
            hours/minutes
          </option>
          <option
            value="PnYnM"
          >
            years/months
          </option>
        </select>
      </div>
    </div>
    <div>
      <div
        class="response-format-datatype-duree-minimum"
      >
        <div
          class="response-format-datatype-duree-label-minimum"
        >
          Minimum
        </div>
        <div
          style="width: 75%;"
        >
          <div
            class="response-format-datatype-duree-minimum"
          >
            <div
              class="ctrl-input "
            >
              <label
                for="input-DURATION.mihours"
              >
                hours
                <span
                  class="ctrl-required"
                >
                  *
                </span>
              </label>
              <div>
                <input
                  id="input-DURATION.mihours"
                  name="DURATION.mihours"
                  placeholder="hours"
                  step="any"
                  type="number"
                  value=""
                />
              </div>
            </div>
            <div
              class="ctrl-input "
            >
              <label
                for="input-DURATION.miminutes"
              >
                minutes
                <span
                  class="ctrl-required"
                >
                  *
                </span>
              </label>
              <div>
                <input
                  id="input-DURATION.miminutes"
                  name="DURATION.miminutes"
                  placeholder="minutes"
                  step="any"
                  type="number"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="response-format-datatype-duree-maximum"
      >
        <div
          class="response-format-datatype-duree-label-maximum"
        >
          Maximum
        </div>
        <div
          style="width: 75%;"
        >
          <div
            class="response-format-datatype-duree-maximum"
          >
            <div
              class="ctrl-input "
            >
              <label
                for="input-DURATION.mahours"
              >
                hours
                <span
                  class="ctrl-required"
                >
                  *
                </span>
              </label>
              <div>
                <input
                  id="input-DURATION.mahours"
                  name="DURATION.mahours"
                  placeholder="hours"
                  step="any"
                  type="number"
                  value=""
                />
              </div>
            </div>
            <div
              class="ctrl-input "
            >
              <label
                for="input-DURATION.maminutes"
              >
                minutes
                <span
                  class="ctrl-required"
                >
                  *
                </span>
              </label>
              <div>
                <input
                  id="input-DURATION.maminutes"
                  name="DURATION.maminutes"
                  placeholder="minutes"
                  step="any"
                  type="number"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Visualize Dropdown Component:  > Should return the right HTML according to format "PTnHnM"  1`] = `
<div>
  <div
    class="response-format-datatype-duree"
  >
    <div
      class="ctrl-select"
    >
      <label
        for="select-DURATION.format"
      >
        Format
        <span
          class="ctrl-required"
        >
          *
        </span>
      </label>
      <div>
        <select
          id="select-DURATION.format"
          name="DURATION.format"
          placeholder="Format"
        >
          <option
            value=""
          >
            Duration format
          </option>
          <option
            value="PTnHnM"
          >
            hours/minutes
          </option>
          <option
            value="PnYnM"
          >
            years/months
          </option>
        </select>
      </div>
    </div>
    <div>
      <div
        class="response-format-datatype-duree-minimum"
      >
        <div
          class="response-format-datatype-duree-label-minimum"
        >
          Minimum
        </div>
        <div
          style="width: 75%;"
        >
          <div
            class="response-format-datatype-duree-minimum"
          >
            <div
              class="ctrl-input "
            >
              <label
                for="input-DURATION.mihours"
              >
                hours
                <span
                  class="ctrl-required"
                >
                  *
                </span>
              </label>
              <div>
                <input
                  id="input-DURATION.mihours"
                  name="DURATION.mihours"
                  placeholder="hours"
                  step="any"
                  type="number"
                  value=""
                />
              </div>
            </div>
            <div
              class="ctrl-input "
            >
              <label
                for="input-DURATION.miminutes"
              >
                minutes
                <span
                  class="ctrl-required"
                >
                  *
                </span>
              </label>
              <div>
                <input
                  id="input-DURATION.miminutes"
                  name="DURATION.miminutes"
                  placeholder="minutes"
                  step="any"
                  type="number"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="response-format-datatype-duree-maximum"
      >
        <div
          class="response-format-datatype-duree-label-maximum"
        >
          Maximum
        </div>
        <div
          style="width: 75%;"
        >
          <div
            class="response-format-datatype-duree-maximum"
          >
            <div
              class="ctrl-input "
            >
              <label
                for="input-DURATION.mahours"
              >
                hours
                <span
                  class="ctrl-required"
                >
                  *
                </span>
              </label>
              <div>
                <input
                  id="input-DURATION.mahours"
                  name="DURATION.mahours"
                  placeholder="hours"
                  step="any"
                  type="number"
                  value=""
                />
              </div>
            </div>
            <div
              class="ctrl-input "
            >
              <label
                for="input-DURATION.maminutes"
              >
                minutes
                <span
                  class="ctrl-required"
                >
                  *
                </span>
              </label>
              <div>
                <input
                  id="input-DURATION.maminutes"
                  name="DURATION.maminutes"
                  placeholder="minutes"
                  step="any"
                  type="number"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Visualize Dropdown Component:  > Should return the right HTML according to format "PnYnM"  1`] = `
<div>
  <div
    class="response-format-datatype-duree"
  >
    <div
      class="ctrl-select"
    >
      <label
        for="select-DURATION.format"
      >
        Format
        <span
          class="ctrl-required"
        >
          *
        </span>
      </label>
      <div>
        <select
          id="select-DURATION.format"
          name="DURATION.format"
          placeholder="Format"
        >
          <option
            value=""
          >
            Duration format
          </option>
          <option
            value="PTnHnM"
          >
            hours/minutes
          </option>
          <option
            value="PnYnM"
          >
            years/months
          </option>
        </select>
      </div>
    </div>
    <div>
      <div
        class="response-format-datatype-duree-minimum"
      >
        <div
          class="response-format-datatype-duree-label-minimum"
        >
          Minimum
        </div>
        <div
          style="width: 75%;"
        >
          <div
            class="response-format-datatype-duree-minimum"
          >
            <div
              class="ctrl-input "
            >
              <label
                for="input-DURATION.miyears"
              >
                years
                <span
                  class="ctrl-required"
                >
                  *
                </span>
              </label>
              <div>
                <input
                  id="input-DURATION.miyears"
                  name="DURATION.miyears"
                  placeholder="years"
                  step="any"
                  type="number"
                  value=""
                />
              </div>
            </div>
            <div
              class="ctrl-input "
            >
              <label
                for="input-DURATION.mimonths"
              >
                months
                <span
                  class="ctrl-required"
                >
                  *
                </span>
              </label>
              <div>
                <input
                  id="input-DURATION.mimonths"
                  name="DURATION.mimonths"
                  placeholder="months"
                  step="any"
                  type="number"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="response-format-datatype-duree-maximum"
      >
        <div
          class="response-format-datatype-duree-label-maximum"
        >
          Maximum
        </div>
        <div
          style="width: 75%;"
        >
          <div
            class="response-format-datatype-duree-maximum"
          >
            <div
              class="ctrl-input "
            >
              <label
                for="input-DURATION.mayears"
              >
                years
                <span
                  class="ctrl-required"
                >
                  *
                </span>
              </label>
              <div>
                <input
                  id="input-DURATION.mayears"
                  name="DURATION.mayears"
                  placeholder="years"
                  step="any"
                  type="number"
                  value=""
                />
              </div>
            </div>
            <div
              class="ctrl-input "
            >
              <label
                for="input-DURATION.mamonths"
              >
                months
                <span
                  class="ctrl-required"
                >
                  *
                </span>
              </label>
              <div>
                <input
                  id="input-DURATION.mamonths"
                  name="DURATION.mamonths"
                  placeholder="months"
                  step="any"
                  type="number"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
